packages/import/src/helpers/labelManager.ts

```diff
@@ -0,0 +1,283 @@
+        this.addLabel({ label: group, teamId });
+      } else if (labelName && existingLabel.id) {
+        const parent = await existingLabel.parent;
+
+        if (parent) {
+          const group = this.idToLabel[parent.id]?.[teamId] as LabelGroup;
+          this.addLabel({ label: new Label(existingLabel.id, labelName), parent: group, teamId });
+        } else {
+          this.addLabel({ label: new Label(existingLabel.id, labelName), teamId });
+        }
+      }
+    }
+  }
+}
+
+const createLabel = async (
+  client: LinearClient,
+  {
+    name,
+    description,
+    color,
+    parentId,
+    teamId,
+  }: {
+    name: string;
+    description?: string;
+    color?: string;
+    parentId?: Id;
+    teamId: Id;
+  }
+) => {
+  const response = await client.createIssueLabel({ name, description, color, teamId, parentId });
+  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+  return (await response?.issueLabel)!.id;
+};
+
+const deleteLabel = async (client: LinearClient, labelId: Id) => {
+  await client.deleteIssueLabel(labelId);
+};
+
+export const handleLabels = async (
+  client: LinearClient,
+  importData: ImportResult,
+  teamId: string,
+  existingLabels: IssueLabel[]
+): Promise<Record<string, { type: "root" | "parent" | "child"; id: Id }>> => {
+  const manager = new LabelManager(existingLabels, teamId);
+  const labelMapping: Record<string, { type: "root" | "parent" | "child"; id: Id }> = {};
+
+  for (const issue of importData.issues) {
+    const labels = issue.labels;
+    if (!labels) {
+      continue;
+    }
+
+    await handleIssueLabels(client, manager, teamId, importData, labels, labelMapping);
+  }
+
+  return labelMapping;
+};
+
+const handleIssueLabels = async (
+  client: LinearClient,
+  manager: LabelManager,
+  teamId: Id,
+  importData: ImportResult,
+  issueLabelIds: string[],
+  labelMapping: Record<string, { type: "root" | "parent" | "child"; id: Id }>
+) => {
+  let actualLabelId: Id | undefined;
+  // Track which groups are used to prevent multiple labels from same group
+  const usedGroups = new Set<string>();
+
+  for (const labelId of issueLabelIds) {
+    const labelData = importData.labels[labelId];
+    const parsed = parseLabelName(labelData.name);
+    const group = parsed[0];
+    let labelName = parsed[1];
+
+    // If this label's group is already used in this issue,
+    // or if it was previously mapped as a child label and its group is in use,
+    // create/use a root label with the full name
+    if (group && (usedGroups.has(group) || (labelMapping[labelId]?.type === "child" && usedGroups.has(group)))) {
+      const fullName = labelData.name;
+      actualLabelId = manager.getLabelByName(fullName, teamId)?.id;
+
+      // If this was previously a child label, delete it before converting to root
+      if (labelMapping[labelId]?.type === "child") {
+        await deleteLabel(client, labelMapping[labelId].id);
+      }
```

Comment:
Is there a risk of conflict here, where we could delete a label that was previously imported and thus may already be applied to some issues?

Reply:
Good spot, there's nothing preventing the case you mentioned. Updated to track whether a label existed before the import in which case that label isn't deleted.